tests/integration/ha/helpers.py

```diff
@@ -0,0 +1,458 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import json
+import logging
+import re
+import subprocess
+from pathlib import Path
+from typing import Dict, Optional
+
+import yaml
+from pytest_operator.plugin import OpsTest
+from tenacity import RetryError, Retrying, retry, stop_after_attempt, wait_fixed
+
+from literals import SERVER_PORT
+
+logger = logging.getLogger(__name__)
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+PROCESS = "/snap/opensearch-dashboards/8/opt/opensearch-dashboards/start.sh"
+SERVICE_DEFAULT_PATH = "/etc/systemd/system/snap.charmed-zookeeper.daemon.service"
+PEER = "cluster"
+
+
+class ProcessError(Exception):
+    """Raised when a process fails."""
+
+
+class ProcessRunningError(Exception):
+    """Raised when a process is running when it is not expected to be."""
+
+
+async def wait_idle(ops_test, apps: list[str] = [APP_NAME], units: int = 3) -> None:
+    """Waits for active/idle on specified application.
+
+    Args:
+        ops_test: OpsTest
+        apps: list of application names to wait for
+        units: integer number of units to wait for for each application
+    """
+    await ops_test.model.wait_for_idle(
+        apps=apps, status="active", timeout=3600, idle_period=30, wait_for_exact_units=units
+    )
+    assert ops_test.model.applications[APP_NAME].status == "active"
+
+
+@retry(
+    wait=wait_fixed(5),
+    stop=stop_after_attempt(60),
+    reraise=True,
+)
+def srvr(host: str) -> dict:
```

Comment:
Yes, I thought I may need the code but I can remove it now :-)

Reply:
In fact I can't -- I'll need (and modify) it in an upcoming task (namely HA process failure).

BTW all this can be removed once _really_ not needed in https://warthogs.atlassian.net/browse/DPE-3954